### PR TITLE
[Hotfix] Create palette

### PR DESCRIPTION
**Pull request by keigohtr**
## What is this PR for?

Since palette is not fixed, label will be invalid for training.
Moreover, prediction process does not use the colormap of training dataset.

This PR fixes these issues.

## This PR includes

- Create a palette when training.
- Create a colormap when predicting.

## What type of PR is it?

Bugfix

## What is the issue?

N/A

## How should this be tested?

N/A

### CODE REVIEW

File: abeja_dataset.py [hunk 1]
```diff
@@ -71,7 +74,7 @@ def __getitem__(self, index):
         # target image
         content = datalake_file.get_content(cache=self.use_cache)
         file_like_object = io.BytesIO(content)
-        target = Image.open(file_like_object).convert('P')
+        target = Image.open(file_like_object).convert('RGB').quantize(palette=self.palette)
```
Comment (keigohtr) [inline abeja_dataset.py:81]: Use `quantize` to apply the same color palette we created.

File: predict.py [hunk 2]
```diff
@@ -15,28 +15,13 @@
...
 import train
 import parameters
+from utils import create_colormap
 
 
 def bitget(number, pos):
     return (number >> pos) & 1
 
 
-def create_colormap(max_num):
```
Comment (keigohtr) [inline predict.py:24]: This colormap is not related to the one of training dataset. I removed and used and alternative.

File: predict.py [hunk 3]
```diff
@@ -48,14 +33,14 @@ def load_model(training_dir, device):
     return model.eval(), num_classes
 
 
-def get_dataset_properties(dataset_ids):
+def get_dataset_labels(dataset_ids):
```
Comment (keigohtr) [inline predict.py:36]: No need to use properties since the current constraint is that we can use only 1 dataset and only 1 category.

File: predict.py [hunk 4]
```diff
@@ -81,8 +66,8 @@ def decode_segmap(out, label_colors):
 device = torch.device(device_name)
 
 model, num_classes = load_model(training_dir, device)
-color_map = create_colormap(num_classes)
-props = get_dataset_properties(dataset_ids)
+dataset_labels = get_dataset_labels(dataset_ids)
+color_map = create_colormap(dataset_labels)
```
Comment (keigohtr) [inline predict.py:71]: This is the alternative of `create_colormap`.

File: utils.py [hunk 5]
```diff
@@ -205,6 +207,24 @@ def log_every(self, iterable, print_freq, header=None):
         print('{} Total time: {}'.format(header, total_time_str))
 
 
+def hex2rgb(hex: str):
+    return tuple(int(hex.lstrip('#')[i:i+2], 16) for i in (0, 2, 4))
```
Comment (keigohtr) [inline utils.py:211]: This is a converter from hex-format string to rgb tuple.

File: utils.py [hunk 6]
```diff
@@ -205,6 +207,24 @@ def log_every(self, iterable, print_freq, header=None):
...
+
+
+def create_colormap(labels: list) -> dict:
+    colormap = {0: (0, 0, 0)}
+    for i, label in enumerate(labels):
+        colormap[i+1] = hex2rgb(label['color'])
+    return colormap
```
Comment (keigohtr) [inline utils.py]: This is the new `create_colormap` which reflects the colormap of training dataset.
index 0 is the background.

File: utils.py [hunk 7]
```diff
@@ -205,6 +207,24 @@ def log_every(self, iterable, print_freq, header=None):
...
+
+
+def create_palette(labels: dict):
+    palette = [[[0, 0, 0]]]
+    for label in labels:
+        palette.append([list(hex2rgb(label['color']))])
+    return Image.fromarray(np.uint8(np.array(palette))).convert("P")
```
Comment (keigohtr) [inline utils.py]: Creating the palette for training.
I created the dummy rgb image which reflects the training dataset to generate the color palette.

**Review by keigohtr — COMMENTED** [review]
I have commented the major changes.

File: utils.py (hunk 6 above)
Comment (contaconta) [reply utils.py]: Is the attribute `label['color']` set by ABEJA Platform Annotation?

Comment (keigohtr) [reply utils.py]: Yes it is.

File: train.py [hunk 8]
```diff
@@ -107,7 +107,7 @@ def get_trainval_dataset_index(dataset_list, early_stopping_test_size):
 def criterion(inputs, target):
     losses = dict()
     for name, x in inputs.items():
-        losses[name] = nn.functional.cross_entropy(x, target, ignore_index=255)
+        losses[name] = nn.functional.cross_entropy(x, target, ignore_index=0)
```
Comment (keigohtr) [inline train.py]: 0 is background

File: train.py [hunk 9]
```diff
@@ -196,6 +196,7 @@ def handler(context):
                 "Training or Test dataset size is too small. "
                 "Please add more dataset or set EARLY_STOPPING_TEST_SIZE properly"
             )
+        drop_last = False if len(dataset) < parameters.BATCH_SIZE else True
```
Comment (keigohtr) [inline train.py:209]: drop_last = False if len(dataset) < parameters.BATCH_SIZE

File: train.py [hunk 10]
```diff
@@ -126,27 +126,33 @@ def evaluate(model, criterion, data_loader, device, num_classes):
...
             confmat.update(target.flatten(), output.argmax(1).flatten())
 
         confmat.reduce_from_all_processes()
 
-    return confmat, total_loss
+    print('eval confmat', confmat)
+    acc_global, _, _ = confmat.compute()
+    epoch_acc = acc_global.item()
 
+    return epoch_acc, sum(epoch_loss)/len(epoch_loss) if len(epoch_loss) else 0.0
```
Comment (keigohtr) [inline train.py:140]: Output accuracy

File: train.py [hunk 11]
```diff
@@ -156,7 +162,11 @@ def train_one_epoch(model, criterion, optimizer, data_loader, lr_scheduler, devi
 
         metric_logger.update(loss=loss.item(), lr=optimizer.param_groups[0]["lr"])
 
-    return total_loss
+    confmat.reduce_from_all_processes()
+    acc_global, _, _ = confmat.compute()
+    epoch_acc = acc_global.item()
+
+    return epoch_acc, sum(epoch_loss)/len(epoch_loss) if len(epoch_loss) else 0.0
```
Comment (keigohtr) [inline train.py:169]: ditto

File: train.py [hunk 12]
```diff
@@ -107,7 +107,7 @@ def get_trainval_dataset_index(dataset_list, early_stopping_test_size):
 def criterion(inputs, target):
     losses = dict()
     for name, x in inputs.items():
-        losses[name] = nn.functional.cross_entropy(x, target, ignore_index=0)
+        losses[name] = nn.functional.cross_entropy(x, target, ignore_index=255)
```
Comment (keigohtr) [inline train.py]: Ignore the padding area

File: transforms.py [hunk 13]
```diff
@@ -58,7 +58,7 @@ def __init__(self, size):
 
     def __call__(self, image, target):
         image = pad_if_smaller(image, self.size)
-        target = pad_if_smaller(target, self.size)
+        target = pad_if_smaller(target, self.size, fill=255)
```
Comment (keigohtr) [inline transforms.py]: Padding area will be ignored by setting them to `255`.
`0` is background area and we want to learn them along with the target label's area.

File: utils.py [hunk 14]
```diff
@@ -238,8 +238,8 @@ def cat_list(images, fill_value=0):
 
 def collate_fn(batch):
     images, targets = list(zip(*batch))
-    batched_imgs = cat_list(images)
-    batched_targets = cat_list(targets)
+    batched_imgs = cat_list(images, fill_value=0)
+    batched_targets = cat_list(targets, fill_value=255)
```
Comment (keigohtr) [inline utils.py]: ditto

**Comment by keigohtr** [comment]
Summary
- Create the fixed color_palette
- Ignore the padding area when calculating the loss
- Ignore `drop_last` on training when `BATCH_SIZE` is bigger than `dataset` size.
- Add `accuracy` score for monitoring.

File: predict.py [hunk 15]
```diff
@@ -48,24 +33,25 @@ def load_model(training_dir, device):
...
-        props = dataset.props
+        labels = dataset.props['categories'][0]['labels']
         break
-    return props
+    return labels
 
 
 def decode_segmap(out, label_colors):
     om = torch.argmax(out.squeeze(), dim=0).detach().cpu().numpy()
+    print(om)
```
Comment (ysku) [inline predict.py]: please remove or add description for `om`

**Review by ysku — APPROVED** [review]
LGTM 👍 for code

File: predict.py (hunk 15 above)
Comment (keigohtr) [reply predict.py]: Adding description might be better because this value is argmaxed prediction result.

Comment (keigohtr) [reply predict.py]: Updated -> 83c2f9a

**Comment by keigohtr** [comment]
Seems no problem. Thank you!

File: train.py (hunk 8 above)
Comment (muchuanyun) [reply train.py]: Sorry, `ignore_index` should still be `255`. The background should also be taken into consideration in loss calculation. (I misunderstood the meaning of `ignore_index`.)
